RInAppMessaging/Classes/Protocols/Lockable.swift

```diff
@@ -18,35 +18,46 @@ internal protocol LockableResource {
 /// Other threads will synchronously wait for unlock() call to continue.
 internal class LockableObject<T>: LockableResource {
     private var resource: T
-    private var lockingThread: Thread?
     private let dispatchGroup = DispatchGroup()
-
-    var isLocked: Bool {
-        return lockingThread != nil
+    @AtomicGetSet private var lockingThread: Thread?
+    @AtomicGetSet private var lockCount: UInt = 0
+    private var shouldWait: Bool {
+        assert(!(isLocked && lockingThread == nil), "Thread was deallocated before calling unlock()")
+        return isLocked && lockingThread != nil && lockingThread != Thread.current
     }
 
+    var isLocked: Bool { lockCount > 0 }
+
     init(_ resource: T) {
         self.resource = resource
     }
 
     deinit {
-        unlock()
+        for _ in [0..<lockCount] {
+            unlock()
```

Comment:
codecov claims the unlock path isn't tested

Reply:
Interesting. I added a breakpoint and it was called multiple times. I think it's because `deinit` isn't called directly in the code.